### PR TITLE
WebGPU: roll header and generated changes in from Dawn

### PR DESCRIPTION
**Pull request by kainino0x**
Rolls to dawn revision:
https://dawn.googlesource.com/dawn/+log/ad853f80af9b236fa3aa6be6dcd2facb26430565
except skips a few entrypoints that can't be implemented in JS (require a C++ shim)
and fixes library_webgpu.js where necessary (function stubs, several of which are going to change before being implemented)

Also adds several files to match Dawn, so that Emscripten mirrors Dawn's generated files directly, requiring less manual intervention, but there is more work to do here.

### CODE REVIEW

**Comment by PZerua** [comment]
I've noticed Chrome expects "float32-filterable" instead of "float32filterable". If I change it manually in library_webgpu.js I can sample float32 textures on web with no problem. Could this be incorporated in this PR?

**Comment by kainino0x** [comment]
> I've noticed Chrome expects "float32-filterable" instead of "float32filterable". If I change it manually in library_webgpu.js I can sample float32 textures on web with no problem. Could this be incorporated in this PR?

Nice catch! Yes, I'll incorporate it into that Dawn change and this PR.

**Comment by kainino0x** [comment]
@beaufortfrancois PTAL

File: system/include/webgpu/README.md [hunk 1]
```diff
@@ -13,27 +13,28 @@ The C++ files are **currently** Dawn-specific, but included in Emscripten for
...
-
-Once that's done, you need to update the auto-generated files with the commands below:
-
-```
-./tools/maint/gen_struct_info.py
-./tools/maint/gen_struct_info.py --wasm64
-./tools/maint/gen_sig_info.py
-```
+To update these bindings from Dawn:
+- Copy `webgpu_cpp_chained_struct.h` from Dawn's source
```
Comment (beaufortfrancois) [inline system/include/webgpu/README.md:17]: Could we link to https://source.chromium.org/chromium/chromium/src/+/main:out/Debug/gen/third_party/dawn/emscripten-bits/system/include/webgpu/webgpu_cpp_chained_struct.h to help readers?

File: system/include/webgpu/README.md [hunk 2]
```diff
@@ -13,27 +13,28 @@ The C++ files are **currently** Dawn-specific, but included in Emscripten for
...
+- Build the `emscripten_bits_gen` target in Dawn (gn or CMake build)
```
Comment (beaufortfrancois) [inline system/include/webgpu/README.md:18]: Note that it also works in Chrome. Maybe rephrase it with 
```suggestion
- Build the `emscripten_bits_gen` Dawn target (gn or CMake build)
```

File: system/include/webgpu/README.md [hunk 3]
```diff
@@ -13,27 +13,28 @@ The C++ files are **currently** Dawn-specific, but included in Emscripten for
...
+- Paste `library_webgpu_enum_tables.js` over the appropriate section of [`library_webgpu.js`](../../../src/library_webgpu.js)
```
Comment (beaufortfrancois) [inline system/include/webgpu/README.md:19]: Could we link to the appropriate section to help?

File: system/include/webgpu/README.md [hunk 4]
```diff
@@ -13,27 +13,28 @@ The C++ files are **currently** Dawn-specific, but included in Emscripten for
...
+- Paste `webgpu_struct_info.json` over the appropriate section of [`struct_info.json`](../../../src/struct_info.json).
```
Comment (beaufortfrancois) [inline system/include/webgpu/README.md:20]: Same here. Could we link to the appropriate section to help?

File: system/include/webgpu/README.md (hunk 3 above)
Comment (beaufortfrancois) [inline system/include/webgpu/README.md:19]: Could we link library_webgpu_enum_tables to https://source.chromium.org/chromium/chromium/src/+/main:out/Debug/gen/third_party/dawn/emscripten-bits/library_webgpu_enum_tables.js?

File: system/include/webgpu/README.md (hunk 4 above)
Comment (beaufortfrancois) [inline system/include/webgpu/README.md:20]: Maybe link to https://source.chromium.org/chromium/chromium/src/+/main:out/Debug/gen/third_party/dawn/emscripten-bits/webgpu_struct_info.json

File: system/include/webgpu/README.md [hunk 5]
```diff
@@ -13,27 +13,28 @@ The C++ files are **currently** Dawn-specific, but included in Emscripten for
...
+- Manually update the `globalThis.gpu` compile-time enum tables (AdapterType, BackendType, etc.)
```
Comment (beaufortfrancois) [inline system/include/webgpu/README.md:21]: Can you elaborate on this. I don't know how this works

File: system/include/webgpu/README.md [hunk 6]
```diff
@@ -13,27 +13,28 @@ The C++ files are **currently** Dawn-specific, but included in Emscripten for
...
-though it changes infrequently.
-
-Once that's done, you need to update the auto-generated files with the commands below:
-
-```
-./tools/maint/gen_struct_info.py
-./tools/maint/gen_struct_info.py --wasm64
-./tools/maint/gen_sig_info.py
-```
+To update these bindings from Dawn:
```
Comment (beaufortfrancois) [inline system/include/webgpu/README.md:16]: We may want to use an ordered list 1,2,3,4 instead of bullet points as they need to be in ordered, especially for the last part where we auto-generate files right?

File: system/include/webgpu/README.md (hunk 1 above)
Comment (kainino0x) [reply system/include/webgpu/README.md:17]: Oops, this was supposed to point to webgpu_enum_class_bitmasks.h. Fixed that, and added a step to copy the `system/` generated files, which I forgot.

File: system/include/webgpu/README.md (hunk 3 above)
Comment (kainino0x) [reply system/include/webgpu/README.md:19]: Not to the line number because it moves, but I can explain this better.
I am hoping to move this into another file so we can remove this manual step, but TBD.

**Comment by kainino0x** [comment]
These instructions are truly out of hand. I will work on this soon :)

I also had to revert some small bits to fix failing compile tests:
> Remove functions that have too-long signatures 
> These functions can't be implemented in JS because the signatures are
too long. They need to be implemented with a C(++) shim over JS but this
will happen in a forthcoming PR.

@beaufortfrancois PTAL

**Comment by kainino0x** [comment]
@sbc100 PTAL as well, I think this is very close to ready to land.

**Comment by kainino0x** [comment]
Oh I just realized François has review privileges here, so maybe that won't be necessary?

File: src/struct_info.json [hunk 7]
```diff
@@ -1227,6 +1227,12 @@
                 "size",
                 "mappedAtCreation"
             ],
+            "WGPUBufferMapCallbackInfo": [
```
Comment (beaufortfrancois) [inline src/struct_info.json:1230]: I assume this should not be there because `MapAsyncF` is not as well?
Does this means all Future related stuff below should be removed also.

File: src/library_sigs.js [hunk 8]
```diff
@@ -1572,6 +1573,7 @@ sigs = {
   wgpuBufferGetSize__sig: 'jp',
   wgpuBufferGetUsage__sig: 'ip',
   wgpuBufferMapAsync__sig: 'vpipppp',
+  wgpuBufferMapAsyncF__sig: 'jpippp',
```
Comment (beaufortfrancois) [inline src/library_sigs.js:1576]: Remove?

File: src/struct_info.json [hunk 9]
```diff
@@ -1268,8 +1274,13 @@
                 "height",
                 "depthOrArrayLayers"
             ],
-            "WGPUInstanceDescriptor": [
-                "nextInChain"
+            "WGPUFuture": [
+                "id"
+            ],
+            "WGPUInstanceFeatures": [
```
Comment (beaufortfrancois) [inline src/struct_info.json:1280]: Is this implemented in Emscripten?

File: system/include/webgpu/README.md [hunk 10]
```diff
@@ -4,32 +4,45 @@ These files, and several snippets of other files, are generated by Dawn
...
+1. Copy [`webgpu_enum_class_bitmasks.h`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/include/webgpu/webgpu_enum_class_bitmasks.h) from Dawn's source
```
Comment (beaufortfrancois) [inline system/include/webgpu/README.md:17]:
```suggestion
1. Copy [`webgpu_enum_class_bitmasks.h`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/include/webgpu/webgpu_enum_class_bitmasks.h) from Dawn's source to `system/include/webgpu/webgpu_enum_class_bitmasks.h`
```

**Review by beaufortfrancois — APPROVED** [review]
LGTM with nits

File: system/lib/webgpu/webgpu_cpp.cpp [hunk 11]
```diff
@@ -1,6 +1,9 @@
+
```
Comment (sbc100) [inline system/lib/webgpu/webgpu_cpp.cpp:1]: Missing copyright header?

**Review by sbc100 — APPROVED** [review]

File: src/struct_info.json (hunk 7 above)
Comment (kainino0x) [reply src/struct_info.json:1230]: I chose not to remove those things adding them back will increase the size of the next roll. So I just removed the minimum needed to make it build (I did this manually, not in dawn.json).

File: src/struct_info.json (hunk 9 above)
Comment (kainino0x) [reply src/struct_info.json:1280]: No, not yet, but it will be soon.

File: src/library_webgpu.js [hunk 12]
```diff
@@ -390,20 +404,20 @@ var LibraryWebGPU = {
       'storage',
       'read-only-storage',
     ],
-    BufferMapState: [
-      'unmapped',
-      'pending',
-      'mapped',
-    ],
+    BufferMapState: {
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply src/library_webgpu.js:407]: Quirk of the code generator, it's because `CompareFunction[0]` is expected to be used ever while `BufferMapState[0]` is not. I was thinking of smoothing this over in the generator along with other generation cleanups (like hopefully moving some generated code into separate files so we can get rid of "please copy-paste" steps)

File: system/include/webgpu/webgpu_cpp_chained_struct.h [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+        ChainedStruct const * nextInChain = nullptr;
+        SType sType = SType(0u);
+    };
+
+    struct ChainedStructOut {
+        ChainedStructOut * nextInChain = nullptr;
+        SType sType = SType(0u);
+    };
+
+}  // namespace wgpu}
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply system/include/webgpu/webgpu_cpp_chained_struct.h:24]: Error in our generator template that had `{{ }}}` instead of `{{ }}`. Good catch :)

File: system/lib/webgpu/webgpu_cpp.cpp [hunk 14]
```diff
@@ -1,6 +1,9 @@
...
+#include <utility>
+
 #include "webgpu/webgpu_cpp.h"
 
-#ifdef __GNUC__
+#if defined(__GNUC__) || defined(__clang__)
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply system/lib/webgpu/webgpu_cpp.cpp:6]: I will propose this change to Dawn first as well.

File: system/lib/webgpu/webgpu_cpp.cpp (hunk 11 above)
Comment (kainino0x) [reply system/lib/webgpu/webgpu_cpp.cpp:1]: Huh, this needs to be fixed upstream. In the meantime let me add the header we have in the template since this one is more important.

Comment (sbc100) [reply system/lib/webgpu/webgpu_cpp.cpp:1]: if it really is from 2017 then this is fine.. I didn't realize webgpu went back that far :)

Comment (kainino0x) [reply system/lib/webgpu/webgpu_cpp.cpp:27]: yup, and this the enum class bitmasks utility probably hasn't even changed much since then. tbd if I update that.

The generated headers will have 2024 in them because the date is written at generation time, so I put those in as 2024.

File: src/library_webgpu.js [hunk 15]
```diff
@@ -662,14 +680,28 @@ var LibraryWebGPU = {
...
+
+  wgpuCreateInstance: (descriptor) => 1,
+
+  wgpuGetInstanceFeatures: () => {
+    abort('TODO: wgpuGetInstanceFeatures unimplemented');
+  },
+
+  wgpuGetProcAddress: (device, procName) => {
+    abort('TODO(#11526): wgpuGetProcAddress unimplemented');
+  },
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply src/library_webgpu.js]: Reverting this change for now because
```
WARNING - [JSC_TYPE_MISMATCH] actual parameter 1 of BigInt does 
not match formal parameter
found   : undefined
required: (bigint|number|string)
  28102|     return BigInt(abort('TODO(#11526): wgpuGetProcAddress unimplemented'));
```
(They're expected to return values and abort doesn't)
